### PR TITLE
Refactor TypeInstrumentation#transformers() method

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
This PR was inspired by DataDog's `AdviceTransformation` interface introduced in https://github.com/DataDog/dd-trace-java/pull/2673 (FYI @tylerbenson)

I've changed the `TypeInstrumentation` interface and introduced `TypeTransformer`:

```java
interface TypeInstrumentation {
  // ...

  // removed the map-returning transformers()
  void transform(TypeTransformer transformer);
}

interface TypeTransformer {

  void applyAdviceToMethod(
      ElementMatcher<? super MethodDescription> methodMatcher, String adviceClassName);

  void applyTransformer(AgentBuilder.Transformer transformer);
}
```

This has several merits:
* we no longer need to create those transformer maps (which I believe was the main reason for the DD change);
* this allows us to introduce type transformations other than advices - e.g. adding the static initializer (see https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2685)

The PR is has lots of changed files, but 98% of it is just changing all existing `TypeInstrumentation`s. The most interesting parts are in `javaagent-extension-api` and `javaagent-tooling`.

### CODE REVIEW

**Comment by laurit** [comment]
If `TypeTransformer` had `applyAdviceToMethod` that takes class instead of className we could build a byte buddy plugin similar to `MuzzleCodeGenerationPlugin` that translates from class to className and calls the string version of `applyAdviceToMethod`.

**Comment by tylerbenson** [comment]
I would suggest splitting this into at least two commits to make it easier to review.

**Comment by mateuszrzeszutek** [comment]
> I would suggest splitting this into at least two commits to make it easier to review.

That's a good idea, done. The first commit in this PR contains all the API changes, the second one is just changing all `TypeInstrumentation`s in the project.

**Review by tylerbenson — APPROVED** [review]
Much better.

File: instrumentation/akka-actor-2.5/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/akkaactor/AkkaActorCellInstrumentation.java [hunk 1]
```diff
@@ -30,8 +28,8 @@
   }
 
   @Override
-  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
```
Comment (trask) [inline instrumentation/akka-actor-2.5/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/akkaactor/AkkaActorCellInstrumentation.java:33]: I will not miss this method signature 😂

**Review by trask — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]
Awesome!

**Pull request merged**